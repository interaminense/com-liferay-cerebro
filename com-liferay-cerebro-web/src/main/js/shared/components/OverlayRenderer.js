import autobind from 'autobind-decorator';
import Component, {Config} from 'metal-jsx';
import dom from 'metal-dom';
import {Align} from 'metal-position';
import {connect} from 'metal-redux';
import {noop} from 'lodash';
import {OrderedMap} from 'immutable';

import Card from 'shared/components/Card';
import overlayRegistry from 'shared/util/overlay-registry';

const OVERLAY_CONTAINER_CLASS = 'overlay-container-root';

class OverlayContainer extends Component {
	getChildContext() {
		const {id, onMouseEnter, onMouseLeave, parentContext} = this.props;

		return {
			overlay: {
				onMouseEnter,
				onMouseLeave,
				parentId: id
			},
			...parentContext
		};
	}

	render() {
		const {children, id, onMouseEnter, onMouseLeave} = this.props;

		return (
			<Card
				data-overlayid={id}
				elementClasses={OVERLAY_CONTAINER_CLASS}
				id={`overlay${id}Container`}
				onMouseEnter={onMouseEnter}
				onMouseLeave={onMouseLeave}
			>
				{children[0]}
			</Card>
		);
	}
}

OverlayContainer.PROPS = {
	id: Config.string().required(),
	onMouseEnter: Config.func().required(),
	onMouseLeave: Config.func().required(),
	parentContext: Config.object().required()
};

class OverlayRenderer extends Component {
	rendered() {
		this.alignOverlays();

		this.toggleBodyListener();
	}

	alignOverlays() {
		this.props.overlaysIOMap.valueSeq().forEach((overlayIMap, i) => {
			const target = document.getElementById(
				`overlay${overlayIMap.get('id')}`
			);

			if (target) {
				const node = this.element.childNodes[i];

				const position = Align.align(
					node,
					target,
					overlayIMap.get('alignment'),
					!overlayIMap.get('forceAlignment')
				);

				this.setOffset(node, position, overlayIMap.get('offset'));
			}
		});
	}

	@autobind
	handleBodyClick(event) {
		const {overlaysIOMap} = this.props;

		const containerNode = dom.closest(
			event.target,
			`.${OVERLAY_CONTAINER_CLASS}`
		);

		const clicked = new Set();

		if (containerNode) {
			let id = containerNode.dataset.overlayid;
			let overlayIMap;

			while ((overlayIMap = overlaysIOMap.get(id))) {
				clicked.add(overlayIMap.get('id'));

				id = overlayIMap.get('parentId');
			}
		}

		overlaysIOMap.forEach(overlayIMap => {
			if (!clicked.has(overlayIMap.get('id'))) {
				overlayIMap.get('onOutsideClick', noop)(event);
			}
		});
	}

	renderOverlay(overlayIMap) {
		const id = overlayIMap.get('id');

		return (
			<OverlayContainer
				elementClasses={overlayIMap.get('containerClass', '')}
				id={id}
				key={id}
				onMouseEnter={overlayIMap.get('onMouseEnter')}
				onMouseLeave={overlayIMap.get('onMouseLeave')}
				parentContext={overlayIMap.get('context')}
			>
				{overlayRegistry.get(id)}
			</OverlayContainer>
		);
	}

	setOffset(node, position, offset) {
		switch (position) {
			case Align.BottomCenter:
			case Align.BottomLeft:
			case Align.BottomRight:
				node.style.transform = `translateY(${offset}px)`;
				break;
			case Align.LeftCenter:
				node.style.transform = `translateX(-${offset}px)`;
				break;
			case Align.RightCenter:
				node.style.transform = `translateX(${offset}px)`;
				break;
			case Align.TopCenter:
			case Align.TopLeft:
			case Align.TopRight:
				node.style.transform = `translateY(-${offset}px)`;
		}
	}

	toggleBodyListener() {
		const active = !this.props.overlaysIOMap.isEmpty();

		if (active && !this._bodyHandler) {
			this._bodyHandler = dom.on(
				document.body,
				'mousedown',
				this.handleBodyClick
			);
		} else if (!active && this._bodyHandler) {
			this._bodyHandler.removeListener();

			this._bodyHandler = null;
		}
	}

	render() {
		const {overlaysIOMap} = this.props;

		return (
			<div class="overlay-renderer-root">
				{overlaysIOMap.map(this.renderOverlay).toArray()}
			</div>
		);
	}
}

const STORE = {
	overlaysIOMap: Config.instanceOf(OrderedMap).required()
};

OverlayRenderer.PROPS = {
	...STORE
};

export default connect(state => ({
	overlaysIOMap: state.get('overlays', OrderedMap())
}))(OverlayRenderer);