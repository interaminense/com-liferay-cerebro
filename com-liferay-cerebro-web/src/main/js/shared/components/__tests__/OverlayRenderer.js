import {Map, OrderedMap} from 'immutable';

import mockStore from 'test/mock-store';
import overlayRegistry from 'shared/util/overlay-registry';
import OverlayRenderer from '../OverlayRenderer';

describe('OverlayRenderer', () => {
	let component;

	afterEach(() => {
		if (component) {
			component.dispose();
		}

		overlayRegistry.clear();
	});

	it('should render', () => {
		component = new OverlayRenderer({
			store: mockStore()
		});

		expect(component).toMatchSnapshot();
	});

	it('should render overlays from the store', () => {
		overlayRegistry.add('1', <div>{'foobar'}</div>);

		component = new OverlayRenderer({
			store: mockStore(
				Map({
					overlays: OrderedMap({
						1: Map({
							context: {},
							id: '1',
							onMouseEnter: jest.fn(),
							onMouseLeave: jest.fn()
						})
					})
				})
			)
		});

		expect(component).toMatchSnapshot();
	});

	it('should render multiple overlays from the store', () => {
		overlayRegistry.add('1', <div>{'overlay1'}</div>);
		overlayRegistry.add('2', <div>{'overlay2'}</div>);

		component = new OverlayRenderer({
			store: mockStore(
				Map({
					overlays: OrderedMap({
						1: Map({
							content: <div>{'overlay1'}</div>,
							context: {},
							id: '1',
							onMouseEnter: jest.fn(),
							onMouseLeave: jest.fn()
						}),
						2: Map({
							content: <div>{'overlay2'}</div>,
							context: {},
							id: '2',
							onMouseEnter: jest.fn(),
							onMouseLeave: jest.fn()
						})
					})
				})
			)
		});

		expect(component).toMatchSnapshot();
	});
});