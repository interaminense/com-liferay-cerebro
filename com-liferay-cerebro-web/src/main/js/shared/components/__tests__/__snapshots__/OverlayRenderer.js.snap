// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverlayRenderer should render 1`] = `<div class="overlay-renderer-root"></div>`;

exports[`OverlayRenderer should render multiple overlays from the store 1`] = `
<div class="overlay-renderer-root">
  <div data-overlayid="1" id="overlay1Container" class="card overlay-container-root">
    <div>overlay1</div>
  </div>
  <div data-overlayid="2" id="overlay2Container" class="card overlay-container-root">
    <div>overlay2</div>
  </div>
</div>
`;

exports[`OverlayRenderer should render overlays from the store 1`] = `
<div class="overlay-renderer-root">
  <div data-overlayid="1" id="overlay1Container" class="card overlay-container-root">
    <div>foobar</div>
  </div>
</div>
`;
